[1-2단계 - 방탈출 예약 대기] 리버(이연후) 미션 제출합니다.

yeonnhuu opened the pull request:
안녕하세요 아서
이번 미션 리뷰 잘 부탁드립니다 😊
감사합니다!

## 체크 리스트
- [x] 미션의 필수 요구사항을 모두 구현했나요?
- [x] Gradle `test`를 실행했을 때, 모든 테스트가 정상적으로 통과했나요?
- [x] 애플리케이션이 정상적으로 실행되나요?
- [x] [프롤로그](https://prolog.techcourse.co.kr)에 셀프 체크를 작성했나요?
- https://prolog.techcourse.co.kr/studylogs/4392

## 객체지향 생활체조 요구사항을 얼마나 잘 충족했다고 생각하시나요?
### 1~5점 중에서 선택해주세요.

- [ ] 1 (전혀 충족하지 못함)
- [ ] 2
- [x] 3 (보통)
- [ ] 4
- [ ] 5 (완벽하게 충족)

### 선택한 점수의 이유를 적어주세요.
1. 한 메서드에 오직 한 단계의 들여쓰기만 한다. ✅
2. else 키워드를 쓰지 않는다. ✅
3. 모든 원시값과 문자열을 포장(wrap)한다. ⚠️
4. 한 줄에 점을 하나만 찍는다. ❌
5. 줄여쓰지 않는다. ⚠️
6. 모든 entity를 작게 유지한다. ✅
7. 2개 이상의 인스턴스 변수를 가진 클래스를 쓰지 않는다. ❌
8. 일급 컬렉션을 쓴다. ❌
9. getter/setter/property를 쓰지 않는다. ❌

## 베이스 코드 선택 체크
- [ ] 이전 미션의 내 코드에서 시작 
- [x] 이전 미션의 페어의 코드에서 시작

## 어떤 부분에 집중하여 리뷰해야 할까요?
각 계층 분리 및 JPA 전환이 잘 되었는지를 중심으로 리뷰해주시면 감사하겠습니다!
[이번 미션에서의 변경사항](https://github.com/woowacourse/spring-roomescape-waiting/pull/244/files/12e1c2f5009c5766bdd0c5f99b0b8202610d2f57..5ddb31a45f08497c8220ee8747219e5d663bae57) 입니다.

### 📌 코드리뷰 질문 그라운드룰 (지우지 마세요)
✅ 많은 사람과 공유하고 토론해보는 것이 더 좋은 질문에는, 리뷰어가 💬말풍선 이모지를 남겨 토론을 추천할 수 있어요.  
✅ 토론을 추천받은 질문은 `#be-7기-잡담-질답`슬랙채널이나 캠퍼스에서 자유롭게 토론을 이어가보세요!  
✅ 코드에 대한 맥락 없이 리뷰어의 주관적인 의견을 묻거나, 단순히 현업에서는 어떻게 하는지를 묻는 질문은 코드 리뷰 본문에는 어울리지 않아요.  
코드 리뷰는 변경된 코드의 목적, 의도, 품질을 중심으로 논의하는 자리예요.  
✅ 코드와 관련된 질문이 있다면 PR 본문에 적기보다는, 해당 코드를 선택해 코멘트를 남겨주세요.

kjyyjk committed:
feat: migrate previous mission code

kjyyjk committed:
chore: jpa 기본 의존성 추가 및 gradle 설정

- DDL 이후 스크립트 실행을 위해 spring.jpa.defer-datasource-initialization=true 설정

Co-authored-by: yeonnhuu <yeonnhuu@gmail.com>

kjyyjk committed:
refactor: 엔티티 연관관계 매핑

Co-authored-by: yeonnhuu <yeonnhuu@gmail.com>

kjyyjk committed:
refactor: Dao를 JpaRepository로 대체

- 추가로 의존성 역전을 통해 Service에서 Jpa 의존성을 제거했다.
덕분에 추후 Jpa가 아닌 다른 기술로 변경할 경우 Repository 구현체만 변경하면 된다.
또한 테스트 시 Fake 객체를 사용할 수 있다.

Co-authored-by: yeonnhuu <yeonnhuu@gmail.com>

yeonnhuu committed:
feat: 내 예약 목록 페이지 구현

Co-authored-by: kjyyjk <rlawnsdud920@gmail.com>

yeonnhuu committed:
feat: 내 예약 목록 API 구현

Co-authored-by: kjyyjk <rlawnsdud920@gmail.com>

kjyyjk committed:
refactor: 예약 상태 표시 리팩토링

Co-authored-by: yeonnhuu <yeonnhuu@gmail.com>

kjyyjk committed:
refactor: 예약 생성 로직 리팩토링

Co-authored-by: yeonnhuu <yeonnhuu@gmail.com>

kjyyjk committed:
refactor: 내 예약 조회 응답 json 수정

Co-authored-by: yeonnhuu <yeonnhuu@gmail.com>

yeonnhuu committed:
refactor: 예약 검색 기능 리팩토링

Co-authored-by: kjyyjk <rlawnsdud920@gmail.com>

kjyyjk committed:
refactor: 클래스 final 키워드 제거

Co-authored-by: yeonnhuu <yeonnhuu@gmail.com>

yeonnhuu committed:
refactor: 예약 검색 기능에서 `@Query` 제거하고 Specification 기반 필터링으로 변경

Co-authored-by: kjyyjk <rlawnsdud920@gmail.com>

yeonnhuu committed:
:recycle: refactor: final 키워드 추가

yeonnhuu committed:
:recycle: refactor: 패키지 구조 개선

yeonnhuu committed:
:recycle: refactor: SQL 스크립트 가독성 개선

Hyunta commented on src/main/java/roomescape/common/error/MissingLoginException.java:
```diff
@@ -1,4 +1,4 @@
-package roomescape.common.exception;
+package roomescape.common.error;
```
패키지 이름을 error라고 바꾸신 이유가 있을까요?

yeonnhuu replied:
해당 패키지는 단순히 예외 클래스만 포함하는 것이 아니라, `ErrorResponse`처럼 에러 응답 형식을 정의하고, `GlobalExceptionHandler`를 통해 예외를 처리하는 클래스들도 함께 포함하고 있습니다. 따라서 전체적인 에러 처리 정책을 다루는 책임을 고려했을 때, 패키지명을 `exception`보다는 `error`로 사용하는 것이 그 역할을 더 명확하게 드러낸다고 판단했습니다.

Hyunta replied:
Java에서는 error 와 exception을 어떻게 분리하는지 알아봅시다.

https://inpa.tistory.com/entry/JAVA-%E2%98%95-%EC%97%90%EB%9F%ACError-%EC%99%80-%EC%98%88%EC%99%B8-%ED%81%B4%EB%9E%98%EC%8A%A4Exception-%F0%9F%92%AF-%EC%B4%9D%EC%A0%95%EB%A6%AC

yeonnhuu replied:
보내주신 내용을 보고 아래와 같이 정리해봤습니다. 현재 `error` 패키지에 포함된 클래스들은 개발자가 직접 처리하는 예외(Exception)들이므로, 패키지명을 `exception`으로 변경하는 것이 더 적절한 것 같습니다. 해당 내용을 반영하도록 하겠습니다.

### (오류 발생 시점에 따른) 프로그래밍 에러
- 논리적 에러(logical error): 버그, 실행은 되지만 의도와는 다르게 동작하는 것
- 컴파일 에러(compile-time error): 문법 구문 오류(syntax error), 컴파일시 발생하는 에러
- 런타임 에러(runtime error): 실행시에 발생하는 에러, 프로그래머 확인 필요

### 런타임에 발생할 수 있는 오류
- 에러(error): 개발자 대처X → `OutOfMemoryError`, `StackOverflowError`
- 예외(exception): 개발자 대처O → 예외 처리(`try-catch`)로 대부분 해결 가능

### 예외(Exception)의 컴파일 에러, 런타임 에러
- 컴파일 에러(Compile Exception): Exception 및 하위 클래스, 컴파일 시 발생하는 예외
  - `IOException`, `FileNotFoundException`, `DataFormatException`
- 런타임 에러(RuntimeException): 프로그래머 실수로 발생
  - `IndexOutOfBoundsException`, `NullPointerException`, `ClassCastException`, `ArithmeticException`

### 예외(Exception)의 체크 예외, 언체크 예외
- 체크 예외(Checked Exception): 컴파일 예외 클래스 → 반드시 예외 처리 필요
  - 별도 예외 처리가 없는 경우 컴파일 자체X, 체크 예외가 발생 가능하다면 `try-catch`로 감싸 `throws`로 던져 처리
- 언체크 예외(Unchecked Exception): 런타임 예외 클래스 → 명시적인 예외 처리X
  - 에러가 일어나는 코드가 있고 `try-catch`로 처리하지 않더라도 컴파일 및 실행 가능
- 필요에 따라 체크 예외를 언체크 예외로 변환해 컴파일에게 예외 처리 강제하지 않도록 처리 가능

Hyunta commented on src/main/java/roomescape/member/domain/Member.java:
```diff
@@ -1,11 +1,25 @@
-    private final MemberRole role;
+
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
```
`@GeneratedValue` 는 어떤 기능을 하나요? 어떤 전략들이 있을까요?

yeonnhuu replied:
`@GeneratedValue`는 엔티티의 `@Id`와 함께 사용되어 기본 키 값을 자동으로 생성해주는 어노테이션입니다. 이때 어떤 방식으로 값을 생성할지 지정하기 위해 strategy 속성을 사용할 수 있으며 대표적으로 다음과 같은 전략들이 있습니다.
- `AUTO`: 기본 전략, JPA가 사용하는 데이터베이스에 따라 적절한 전략을 자동으로 선택
- `IDENTITY`: 데이터베이스의 AUTO_INCREMENT 기능을 활용해 기본 키를 생성
- `SEQUENCE`: 데이터베이스 시퀀스를 사용해 기본 키를 생성
- `TABLE`: 별도의 테이블을 생성해 기본 키를 관리

이번 미션에서는 기본 키 생성을 데이터베이스가 자동으로 처리하도록 하기 위해 `IDENTITY` 전략을 사용해 `@GeneratedValue`를 설정했습니다.

Hyunta commented on src/main/java/roomescape/member/domain/Member.java:
```diff
@@ -1,11 +1,25 @@
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+    @Embedded
```
`@Embedded`는 어떤 역할을 하나요?

yeonnhuu replied:
`@Embedded`는 엔티티 내부에 또 다른 객체를 포함된 형태로 매핑할 때 사용합니다. 이 어노테이션을 사용하면 별도의 테이블을 만들지 않아도 포함된 객체의 필드들이 엔티티의 테이블 내 컬럼으로 함께 저장됩니다. `Member` 엔티티에서는 `MemberName`과 `MemberRole` 필드를 각각 `@Embedded`로 선언했으며, 이 두 클래스는 `@Embeddable`로 정의하여 JPA가 포함 객체로 인식할 수 있도록 했습니다.

Hyunta commented on src/main/java/roomescape/member/domain/Member.java:
```diff
@@ -1,11 +1,25 @@
+    private MemberName name;
+    private String email;
+    private String password;
+    @Enumerated(EnumType.STRING)
```
`@Enumerated` 는 어떤 기능을 하나요?

yeonnhuu replied:
`@Enumerated`는 JPA에서 enum 타입 필드를 데이터베이스에 어떤 형식으로 저장할지 지정하는 어노테이션입니다. 기본 방식인 `EnumType.ORDINAL`은 enum의 선언 순서에 따른 숫자를 저장하고, `EnumType.STRING`은 enum의 문자열 이름을 그대로 저장합니다. `ORDINAL` 방식은 enum의 순서가 바뀌거나 값이 추가될 경우 데이터 불일치 위험이 있기 때문에 이번 미션에서는 `MemberRole` 필드를 `EnumType.STRING`으로 명시적으로 설정했습니다.

Hyunta commented on src/main/java/roomescape/member/domain/MemberName.java:
```diff
@@ -1,16 +1,22 @@
-public final class MemberName {
+import jakarta.persistence.Embeddable;
+
+@Embeddable
+public class MemberName {
```
어떤 기준으로 Embeddable 객체를 나누셨나요?

yeonnhuu replied:
유효성 검증이 필요한 값에 대해서는 별도의 `@Embeddable` 클래스로 분리하여 값 객체로 설계했습니다. 예를 들어 `MemberName`의 경우 이름은 최대 5글자 이내여야 한다는 비즈니스 규칙이 존재하므로, 해당 로직을 캡슐화하기 위해 별도의 값 객체로 분리했습니다. 반면 현재 `email`과 `password`는 별도의 도메인 유효성 검증이나 비즈니스 규칙이 없어 일반 `String` 필드로 유지하고 있습니다. 다만 향후 유효성 검증이 필요해진다면 `@Embeddable` 클래스로 분리하여 관리할 예정입니다.

Hyunta commented on src/main/java/roomescape/member/infrastructure/jpa/MemberRepositoryImpl.java:
```diff
@@ -0,0 +1,37 @@
+import roomescape.member.domain.Member;
+import roomescape.member.domain.MemberRepository;
+
+@Repository
+public class MemberRepositoryImpl implements MemberRepository {
+
+    private final MemberJpaRepository memberJpaRepository;
```
Repository가 Repository를 가지고 있는 구조로 만들어주셨네요.
왜 이런 구조를 선택하셨을까요?

yeonnhuu replied:
계층 간 결합도를 낮춰 유연한 구조를 만들기 위해 Repository Adapter 패턴을 도입했습니다.

현재 `MemberRepository`는 도메인 계층에서 사용하는 인터페이스고, `MemberRepositoryImpl`은 이를 실제로 구현한 인프라스트럭처 계층의 클래스입니다. 이렇게 인터페이스와 구현을 분리하면, 도메인 로직이 JPA 같은 특정 기술에 직접 의존하지 않아도 되기 때문에 확장성 측면에서 유리합니다.

또한 이 구조는 의존성 역전 원칙(DIP)과 인터페이스 분리 원칙(ISP)을 따릅니다. 도메인 계층은 구현체가 아니라 인터페이스에만 의존하고 실제 구현은 인프라 계층에서 맡게 되어 DIP를 만족합니다. `JpaRepository`처럼 기능이 많은 인터페이스를 그대로 사용하는 대신 도메인에서는 필요한 기능만 담은 인터페이스(`MemberRepository`)를 따로 정의해서 ISP를 충족합니다.

이런 구조를 사용하면 나중에 JPA가 아닌 다른 기술을 도입하더라도 `MemberRepositoryImpl`만 교체하면 되고, 도메인 로직은 그대로 둘 수 있어서 기술 변경에도 유연합니다. 또한 테스트할 때는 실제 DB를 사용하는 `MemberJpaRepository `대신, Fake나 Mock 객체를 써서 빠르고 독립적인 단위 테스트를 작성할 수 있다는 장점도 있습니다.

Hyunta replied:
작성해주신 내용 중에 잘못된 내용은 없습니다, 그런데 개인적으로는 확장성을 지나치게 고려하는 것이 아닐까 하는 생각도 듭니다.
대신에 SpringDataJPA 에서 자동으로 메서드를 구현해주는 기능을 사용하지 못하게 됩니다. 빈도를 생각해봤을 데이터 접근 방식을 변경하는 일은 드물게 일어날 것 같아요. 그럼에도 불구하고 확장성있는 구조가 더 유리하다고 생각하실까요?
평소에 SpringDataJPA를 사용하면서 생산성을 높게 유지하다가, 데이터 접근 기술을 변경할 때 시간을 조금 투자하는 편이 더 효율적이지 않을까요?

yeonnhuu replied:
말씀하신 내용에 공감합니다! 저도 실제로 Spring Data JPA의 편리한 기능들을 제대로 활용하지 못했던 점이 아쉬웠는데, 처음부터 JPA를 사용했다면 Repository Adapter 구조까지는 생각하지 않았을 것 같습니다. JDBC에서 JPA로 전환하면서 확장성에 필요 이상으로 치중해서 고민했던 것 같고, 말씀처럼 생산성을 먼저 챙기고 필요할 때 리팩터링하는 게 더 현실적일 것 같습니다.

Hyunta commented on src/main/java/roomescape/member/infrastructure/jpa/MemberJpaRepository.java:
```diff
@@ -0,0 +1,10 @@
+package roomescape.member.infrastructure.jpa;
```
Repository의 패키지를 왜 infrastructure 하위에 두셨나요?

yeonnhuu replied:
domain은 비즈니스 로직을 담당하고 infrastructure는 외부 시스템과의 연결을 포함한 기술적인 구현을 담당하도록 계층 간 책임을 명확히 분리하고자 infrastructure 패키지를 별도로 구성했습니다. `MemberJpaRepository`는 Spring Data JPA에 의존하는 구체적인 기술 구현체로, 비즈니스 로직과는 직접적인 관련이 없다고 판단해 infrastructure 패키지 하위에 위치시켰습니다.

이처럼 기술 구현을 인프라 계층에 한정하면 도메인 계층이 특정 기술에 의존하지 않도록 유지할 수 있습니다. 또한 추후에 JPA가 아닌 다른 데이터 접근 기술로 전환이 필요할 경우, 변경 범위를 인프라 계층 내부로 국한할 수 있어 확장성과 유지보수성 측면에서도 유리하다고 생각했습니다.

Hyunta replied:
해당 내용은 비즈니스 로직 아닌가요?

``` java
public interface ThemeJpaRepository extends JpaRepository<Theme, Long> {

    boolean existsByName(ThemeName name);

    @Query("""
            SELECT r.theme, t.name, t.thumbnail, t.description, COUNT(r.id) AS cnt
            FROM Reservation r
            JOIN Theme t ON t = r.theme
            WHERE r.date BETWEEN :from AND :to
            GROUP BY r.theme
            ORDER BY cnt DESC
            LIMIT :count
            """)
    List<Theme> findPopularThemes(@Param("from") LocalDate from, @Param("to") LocalDate to, @Param("count") int count);
}
```

yeonnhuu replied:
이전 미션에서 사용한 쿼리를 그대로 가져오면서 이 부분의 리팩터링을 놓친 것 같습니다. Specifications를 활용해서 인프라 계층에 비즈니스 로직이 들어가지 않도록 개선하겠습니다.

Hyunta commented on src/main/java/roomescape/reservation/domain/ReservationRepository.java:
```diff
@@ -0,0 +1,25 @@
+import java.time.LocalDate;
+import java.util.List;
+
+public interface ReservationRepository {
+
+    List<Reservation> findAll();
+
+    Reservation save(Reservation reservation);
+
+    void deleteById(long id);
+
+    boolean existsByDateAndTimeIdAndThemeId(LocalDate date, long timeId, long themeId);
+
+    boolean existsByTimeId(long timeId);
+
+    boolean existsByThemeId(long themeId);
+
+    List<Reservation> findAllByDateAndThemeId(LocalDate date, long themeId);
+
+    List<Reservation> findAllByMemberId(long id);
+
+    List<Reservation> findAllByCondition(Long memberId, Long themeId, LocalDate from, LocalDate to);
+}
```
인터페이스를 2개를 놓아서 매번 필요한 메서드를 정의해줘야 사용할 수 있는 구조가 되었네요.
왜 이렇게 하는게 더 낫다고 생각하셨을까요?

yeonnhuu replied:
JPA나 특정 기술에 도메인 계층이 직접 의존하지 않도록 설계하기 위해 `ReservationRepository`처럼 별도의 도메인 인터페이스를 두고 필요한 메서드를 정의했습니다. Spring Data JPA가 제공하는 `JpaRepository`는 도메인 입장에서는 필요 없는 메서드까지 그대로 노출되기 때문에 이는 곧 인터페이스 분리 원칙(ISP)을 위반 합니다. 그래서 도메인 계층에서는 `ReservationRepository`처럼 업무에 필요한 메서드만 추상화한 인터페이스를 두고, 실제 구현체(`ReservationRepositoryImpl`)에서는 JPA 기반의 `ReservationJpaRepository`에 위임하는 구조를 사용했습니다.

Hyunta reviewed (changes requested):
안녕하세요 리버,
방탈출 예약 대기 미션 리뷰어로 만나게 된 아서입니다.
미션을 잘 진행해주셨네요.
몇가지 커멘트 남겨드렸는데 확인부탁드립니다.
궁금하거나 모호한 부분이 있으면 DM 주세요

yeonnhuu committed:
:recycle: refactor: `Lombok` 적용 및 `ThemeThumbnail` 분리, 테스트 가독성 개선

- `@Getter`, `@NoArgsConstructor` 등 `Lombok`을 적용해 반복 코드 제거
- `Theme`의 `thumbnai`l 필드를 `ThemeThumbnail` 값 객체로 분리하여 책임 명확화
- 테스트 코드의 가독성을 높이기 위한 구조 개선

Hyunta reviewed (approved):
안녕하세요 리버,
남겨주신 피드백에 커멘트 달아드렸습니다.
추가로 궁금하시거나 모호한 부분이 있으면 DM 주세요.
이번 단계는 여기서 머지하도록 하겠습니다.
